### ISSUE TITLE
package.json is Difficult to Maintain Without Comments

### ISSUE THREAD

**Issue by machineghost** [npm/npm#13654]
I understand that grumpy old man Crockford, father of JSON, didn't believe in JSON comments, and I understand Node originally based `package.json` on Crockford's specification ... but Node isn't Crokford's project, and Crockford's interests and the interests of thousands of Node users are not aligned.

Can you please provide a syntax for comments in `package.json`?  It's not about ideology, it's not about sticking to your gun, it's about serving the people who use Node.  They're human beings, and human beings didn't evolve to read `package.json` out in the savanna.  Now that we're in cubicles we're still just as bad at it.

Please, do what's best for Node users, not what Crockford wants.  Any string parser in the world can extract `/* foo */`.

**Comment by EmmaRamirez** [comment]
I think that there is _some_ importance in following standards, especially for something as widely adopted as JSON.

I think that for the most part, the `README` of a project is a better place for long-term documentation than the `package.json` file itself. I think perhaps I'm not envisioning a good enough scenario to warrant comments within `package.json`, partly because the code there is generally self-documenting (scripts are fairly straightforward and should definitely be mentioned in actual documentation, dependencies don't necessarily seem to need much explanation and change very often anyway if not shrinkwrapped).

Though I would *not* recommend this, [you can include your package.json in your `index.js`](https://github.com/npm/read-package-json#indexjs), though you would not be able to comment within the comment.

**Comment by machineghost** [comment]
>I think that there is some importance in following standards

Absolutely!  But I'd argue that interest isn't as important as building a tool that humans can use well.

>I think that for the most part, the README of a project is a better place for long-term documentation

Amen!  But just as some JS comments are JSDoc and some are short `//` comments, there's value to both in documenting your `package.json`.  Being able to understand it without having to simultaneously read another document, because the config file itself has basic "signpost" comments, is tremendously valuable.

To put it another way, forget about NPM and `package.json` entirely for just a second: if you were creating a configuration file for *any* system, and you knew that file would become complex, would you or would you not implement comments?

**Comment by t1st3** [comment]
@machineghost You could also use something like [`p-s`](https://github.com/kentcdodds/p-s), which basically allows you to write your npm scripts in a separate .js file, in which comments are welcome, and thus circumvent the no-comment-in-json ruling.

**Comment by machineghost** [comment]
>@machineghost You could also use something like p-s, 

Totally.  For me it's a cost/benefit trade-off: using 3rd party tools this way always has a cost.  In my experience, it's often a non-obvious cost that comes back to bite you at the worse possible time :)  This is why, even though I <3 Jeremy Ashkenas (that man is a genius) I would never use CoffeeScript.

Similarly here, if my only options are to incur that cost or just suck it up, I'm going to just suck it up.  I'm just hoping that:
A) I'm not alone in this dilemma
B) the people on this project truly want to server their users
C) they will see that I'm not alone, and that this tickiet is an opportunity to better serve those users

**Comment by KenanY** [comment]
Previously discussions: https://github.com/npm/npm/issues/4482, https://github.com/npm/read-package-json/pull/46

**Comment by machineghost** [comment]
Well, it looks like A) is true at least :) ... it's just too bad either B) or C) isn't.

It's a bummer, but sorry for beating what evidently is a dead horse.

**Comment by othiym23** [comment]
While the CLI team and the project do what we do because we want to support you, the users of the CLI, and as much as possible keeping the CLI out of your way as a workflow tool. That said:

1. As alluded to in the previous iterations of this discussion (thanks, @KenanY), there's a large backwards compatibility issue inherent in departing from strict JSON. Even were npm to normalize to strict JSON at publish time, that would break older versions of npm trying to install via Git.
2. npm isn't the only tool that uses `package.json`. Among other things, Node itself reads that file (using `JSON.parse()`) to determine the entry point for a module. This makes changing / relaxing the file format a pretty tough sell, community-wise.
3. As others have mentioned, there are other tools that allow you to translate from other formats (YAML, CSON, pure JS, CoffeeScript) into JSON.

I get the appeal of having a blessed solution and not having to add an intermediate step. It's definitely a better user experience. At the same time, wearing my product / project manager's hat, this is a very expensive change to land, and there are many, much higher priorities for the project to tackle. As such, as we've done in the past, I'm going to close this particular request. Thanks for your time, patience, and understanding.
